Submit Investment data to Data Hub

saruniitr opened the pull request:
## Change description

This PR adds changes to submit investment enquiry data to Data Hub.

During this process it first checks whether a company exists in Data Hub otherwise it returns an error (see below). Next we check for any existing contacts and add the contact as secondary/primary. It also searches for the assigned adviser and finally submits the data.

At any point if there is an error, the list of errors till that point are presented to the user (as per example screenshot below).

The relevant metadata is fetched and mapped to the unique ids as required by investment creation API. There is a mismatch of reference data coming from the website as compared to what is available in Data Hub, in this case we won't get a uuid and treated as error.

SSO is now integrated so there are few changes to remove previous TODOs.

Company not available in Data Hub
![company-not-in-dh](https://user-images.githubusercontent.com/2622464/78323817-e6f4af80-756a-11ea-8cb4-419d3cd9e99b.png)

reupen commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -213,3 +222,205 @@ def dh_contact_search(contact_name):
+        "job_title": enquirer.job_title,
+        "company": company_id,
+        "primary": primary,
+        "telephone_countrycode": "NOT SET",
```
This should be a real value rather than NOT SET. So either the phone number needs to be captured from the user as two fields, or it needs splitting afterwards.

saruniitr replied:
Unfortunately the form where user enters this data has one column, no field to capture the code so we cannot rely on the number to split it up later. We have a backlog item to add a field in the tool then service team can enter the code based on the country. I was planning to do a look up but I think I make the form changes in our tool itself to enter the country code.

reupen commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -213,3 +222,205 @@ def dh_contact_search(contact_name):
+    advisers = []
+    url = f"{settings.DATA_HUB_ADVISER_SEARCH_URL}/?autocomplete={adviser_name}"
+
+    response = dh_request(request, access_token, "GET", url, {})
```
It would be good to pass query params to requests using the `params` keyword argument so that they're escaped correctly.

saruniitr replied:
ok

reupen commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -213,3 +222,205 @@ def dh_contact_search(contact_name):
+                "name": adviser["first_name"],
+                "is_active": adviser["is_active"],
+            }
+        )
```
It would be good to use a list comprehension here e.g.:

```python
advisers = [
    {
        "datahub_id": adviser["id"],
        "name": adviser["first_name"],
        "is_active": adviser["is_active"],
    }
    for adviser in response.json()["results"]
]
```

It's both idiomatic and more efficient.

saruniitr replied:
I was favouring readability but I can change it to use list comprehension.

reupen commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -213,3 +222,205 @@ def dh_contact_search(contact_name):
+        return response
+
+    # check if the user is available in Data Hub
+    advisers, error = dh_adviser_search(request, access_token, request.user)
```
Is this trying to get the adviser associated with the access token? The usual way to do this is to make a request to `/whoami/`.

elcct replied:
If you are trying to find user by name, be aware that there are cases where advisers share the same name. I think what @reupen  suggests is the way to go.

saruniitr replied:
Yes this is find the user on Data Hub as they will be recorded as referral source adviser. I will update this to make use of `/whoami/`

reupen commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -213,3 +222,205 @@ def dh_contact_search(contact_name):
+    )
+
+    # It is always default value - corresponds to Services
+    payload["business_activities"] = ["2f51ea6a-ca2f-466a-87fd-5f79ebfec125"]
```
It would be good to make this UUID a named constant.

saruniitr replied:
agree. Initially I was thinking of importing data hub fixtures and load it up from there but it can quickly go out of sync and adds maintenance. Since this is a fixed field I have used the id directly but will make it named for readability.

reupen commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -213,3 +222,205 @@ def dh_contact_search(contact_name):
+    )
+    payload["referral_source_activity_website"] = get_dh_id(
+        dh_metadata["referral-source-website"], "Invest in GREAT Britain"
+    )
```
It would be good to be consistent about whether these are referenced by name or by ID. I would suggest ID (possibly even overridable in setting) as they can be freely renamed from the admin site while the UUIDs are fairly persistent.

saruniitr replied:
ok, there are only few fields that require this. Plan was to refine this in later releases.

saruniitr replied:
Also in future these may come from the enquiry model itself, for now I have changed them as defines.

reupen commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -213,3 +222,205 @@ def dh_contact_search(contact_name):
+    return item[0]["id"]
+
+
+def dh_investment_create(request, enquiry, metadata=None):
```
This function is quite large – it would be good to split it up.

saruniitr replied:
I mainly included in a single function to handle errors easily and to collect error descriptions as we need to display them in the UI. I will see if it can be refactored.

reupen commented on app/settings/common.py:
```diff
@@ -49,7 +49,7 @@
     'loggers': {
         '': {
-            'level': 'DEBUG' if env.bool('DEBUG') else 'INFO',
+            'level': 'INFO' if env.bool('DEBUG') else 'INFO',
```
Can this be just 'INFO' now?
```suggestion
            'level': 'INFO',
```

saruniitr replied:
yes, finding DEBUG too verbose, will make it INFO.

reupen commented on app/enquiries/tests/test_dh_integration.py:
```diff
@@ -161,6 +174,41 @@ def test_contact_search_error(self):
             self.assertIsNotNone(error)
             self.assertEqual(error["name"], expected["name"])
+
+    def test_investment_creation_fails_company_not_in_dh(self):
```
It there a test for when `dh_investment_create()` succeeds? It would be good to check that the expected data is being sent to Data Hub.

saruniitr replied:
Not at the moment, adding it.

elcct replied:
Has this test been added?

saruniitr replied:
No, sandbox api doesn't support this yet so could not add it.

elcct commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -213,3 +222,205 @@ def dh_contact_search(contact_name):
+    }
+
+    # Extract access token
+    session = get_oauth_payload(request)
```
I think there is a duplication of this logic. Why the token needs to be passed explicitly if it is being extracted in `dh_request` if not provided? It would be good to clean this up.

saruniitr replied:
I am passing both `request` and `access_token` to `dh_request` to have some flexibility as it is used by other functions (company/contact/adviser search, contract create and investment create).
Only investment create receives `request` object which calls other functions but there is a case where company search is called separately. To cover these case I am sending token separately. I agree it causes some duplication but favoured flexibility to support all cases.

elcct commented on app/enquiries/models.py:
```diff
@@ -216,6 +214,40 @@ class Enquiry(TimeStampedModel):
+        null=True,
+        verbose_name="Duns number",
+    )
+    dh_assigned_company = models.CharField(
```
It would be good to rename it to `dh_assigned_company_name`, as otherwise it suggests the column may store the company object, not just the name.

saruniitr replied:
hmm, will change it to make it clear.

saruniitr commented:
@reupen @elcct @mcbhenwood  With reference to tests, my limitation is that I can only mock requests to datahub as the DH API is not running locally. CI cannot really access data hub so no option but to mock responses. This means I need to setup lot of mock responses, data hub metadata in this project and needs to be updated as and when API changes. Do you have any recommendation on how to resolve this?

reupen commented:
I think that is right for unit-ish tests. We don't have any kind of mock client available in Python or any other language, so it would be requests-mock I think.

For end-to-end tests, you could use the data-hub-api Docker image with its fixtures. But I don't believe there are any end-to-end tests in this repo at the moment.

(There is also something called that sandbox in the Data Hub FE, but that's a bunch of fixtures and a fake API server.)

In relation to metadata lists, there are two types – ones that are managed in the Data Hub code base and ones that are editable from the admin site. For the second group, fewer assumptions can be made.

reupen commented:
@saruniitr Let me know when this is ready for re-review.

saruniitr commented:
> @saruniitr Let me know when this is ready for re-review.

Addressed previous comments and refactored it. Once the tests are complete will let you know.

saruniitr commented:
Have address previous review comments.
Also integrated cypress test framework and started adding a test for investment creation using sandbox api but the `POST` request on `/v3/investment` giving `route invalid` error. Hence the changes are not committed to this PR. Checking with FE team about that error, if that is resolved then I will push those changes.

elcct commented on .circleci/config.yml:
```diff
@@ -18,6 +18,10 @@ jobs:
+      DATA_HUB_CONTACT_CREATE_URL: 'https://datahub/contact/create'
+      DATA_HUB_INVESTMENT_CREATE_URL: 'https://datahub/investment/create'
+      DATA_HUB_ADVISER_SEARCH_URL: 'https://adviser/search'
+      DATA_HUB_WHOAMI: 'https://datahub/whoami/'
```
It is missing `_URL`.

saruniitr replied:
fixed

elcct commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -24,7 +28,17 @@
+    url,
+    payload,
+    request_headers=None,
+    params={},
```
It's probably better to have `None` as default for `params` and then assign empty dictionary inside the function.

saruniitr replied:
yes agree, as it is mutable it will cause issues with default arguments, corrected it.

peterhudec commented on app/enquiries/templates/enquiry_detail.html:
```diff
@@ -18,6 +18,27 @@
+            <h2 class="govuk-error-summary__title" id="error-summary-title">
+              Please correct below errors before submitting again,
+            </h2>
+            {% for error in errors %}
+                <div class="govuk-error-summary__body">
+                <ul class="govuk-list govuk-error-summary__list">
+                    {% for key,value in error.items %}
+                    <li> {{ key.capitalize }}: {{ value}} </li>
+                    {% endfor %}
+                </ul>
+                </div>
+            {% endfor %}
```
The outer loop looks weird, there surely should only be a single `.govuk-error-summary__body`. I would expect that the for loop in the `<ul>` would iterate over the errors. What are `error.items`?

saruniitr replied:
Here errors is a dict object, these are not form errors. In the backend a dictionary is constructed with keys as error field and value as description of the error. Please see dh_enquiry_readiness() in datahub_utils.py. An example is shown below,
```
response = {"errors": []}
# Allow creating of investments only if Company exists on DH
if not enquiry.dh_company_id:
    response["errors"].append(
        {"company": f"{enquiry.company_name} doesn't exist in Data Hub"}
...
....
```

web-bert replied:
I agree with @peterhudec also the `<li>` should have a link in it to the element in the error state. If the errors are not in a format that works for the template then they should be normalised to work for the template. I would also argue that the error summary should be a reusable snippet that takes a list of errors with a message and an id to link to the element in the error state

peterhudec commented on app/enquiries/templates/enquiry_detail.html:
```diff
@@ -105,31 +127,38 @@ <h3 class="govuk-heading-m">HPO details</h3>
+    <form class="form" action="{% url 'enquiry-detail' pk=enquiry.id %}" method="post">
+        {% csrf_token %}
+
+        {% if not enquiry.date_added_to_datahub %}
```
Shouldn't the whole form be inside of this condition?

saruniitr replied:
It can be but this is a static page and there are no input fields, just a button at the bottom of a lengthy page to submit, for this reason only enclosed the button in the form tag.

reupen commented on app/enquiries/common/datahub_utils.py:
```diff
@@ -187,31 +221,310 @@ def dh_company_search(company_name):
+    Returns list of subset of fields for each Adviser found
+    """
+    advisers = []
+    url = f"{settings.DATA_HUB_ADVISER_SEARCH_URL}"
```
Looks like this f-string could be removed:

```suggestion
    url = settings.DATA_HUB_ADVISER_SEARCH_URL
```

web-bert commented on app/enquiries/templates/enquiry_detail.html:
```diff
@@ -105,31 +127,38 @@ <h3 class="govuk-heading-m">HPO details</h3>
-</main>
+        <div class="govuk-grid-row enquiry-details-row">
+            <h3 class="govuk-heading-m">Data Hub details</h3>
+            <dl class="govuk-summary-list">
+                {% include 'snippets/summary_list_key_value.html' with instance=enquiry field="organisation_type" %}
+                {% include 'snippets/summary_list_key_value.html' with instance=enquiry field="investment_type" %}
+                {% include 'snippets/summary_list_key_value.html' with instance=enquiry field="project_name" %}
+                {% include 'snippets/summary_list_key_value.html' with instance=enquiry field="project_description" %}
+                {% include 'snippets/summary_list_key_value.html' with instance=enquiry field="anonymised_project_description" %}
+                {% include 'snippets/summary_list_key_value.html' with instance=enquiry field="estimated_land_date" %}
+                {% include 'snippets/summary_list_key_value.html' with instance=enquiry field="new_existing_investor" %}
+                {% include 'snippets/summary_list_key_value.html' with instance=enquiry field="investor_involvement_level" %}
+                {% include 'snippets/summary_list_key_value.html' with instance=enquiry field="specific_investment_programme" %}
+                {% include 'snippets/summary_list_key_value.html' with instance=enquiry field="crm" %}
+                {% include 'snippets/summary_list_key_value.html' with instance=enquiry field="date_added_to_datahub" %}
+                {% include 'snippets/summary_list_key_value.html' with instance=enquiry field="project_code" %}
+                {% include 'snippets/summary_list_key_value.html' with instance=enquiry field="datahub_project_status" %}
+                {% include 'snippets/summary_list_key_value.html' with instance=enquiry field="project_success_date" %}
+            </dl>
```
This could be more DRY by creating a reusable `summary-list` snippet?

saruniitr commented:
Thanks for your reviews, received three approvals, merging it.
